File: src-pyqt/core/ludusavi_manifest_provider.py
"""
Offline Ludusavi save-path manifest: bundled seed, user-data cache, optional online refresh.
"""

from __future__ import annotations

import json
import os
import re
import tempfile
import threading
import time
from dataclasses import dataclass
from enum import Enum
from typing import Dict, List, Optional

import requests
import yaml

from core.user_data_dir import get_pyqt_user_data_dir

MANIFEST_URL = "https://raw.githubusercontent.com/mtkennerly/ludusavi-manifest/master/data/manifest.yaml"
MANIFEST_FILENAME = "ludusavi-save-manifest.json"
META_FILENAME = "ludusavi-save-manifest.meta.json"
MAX_MANIFEST_DOWNLOAD_BYTES = 64 * 1024 * 1024
MANIFEST_HTTP_TIMEOUT_SEC = 180

_NAME_CLEAN = re.compile(r"[^a-z0-9]+", re.IGNORECASE)


class LudusaviMatchKind(str, Enum):
    NONE = "none"
    STEAM_ID = "steam_id"
    NAME_INDEX = "name_index"


@dataclass(frozen=True)
class LudusaviSaveLookup:
    paths: List[str]
    match_kind: LudusaviMatchKind


def normalize_manifest_game_name(name: str) -> str:
    if not name or not str(name).strip():
        return ""
    return _NAME_CLEAN.sub(" ", str(name).strip().lower()).strip()


def default_bundled_manifest_path() -> str:
    return os.path.join(os.path.dirname(os.path.dirname(__file__)), "data", MANIFEST_FILENAME)


def _atomic_write_text(path: str, text: str) -> None:
    directory = os.path.dirname(path)
    os.makedirs(directory, exist_ok=True)
    fd, tmp = tempfile.mkstemp(dir=directory, prefix=".tmp_", suffix=".json")
    try:
        with os.fdopen(fd, "w", encoding="utf-8") as handle:
            handle.write(text)
        os.replace(tmp, path)
    except Exception:
        try:
            os.unlink(tmp)
        except OSError:
            pass
        raise


def _translate_manifest_path(path: str) -> str:
    mapping = {
        "<home>": "~",
        "<winAppData>": "%APPDATA%",
        "<winLocalAppData>": "%LOCALAPPDATA%",
        "<winLocalAppDataLow>": "%USERPROFILE%\\AppData\\LocalLow",
        "<winDocuments>": "%USERPROFILE%\\Documents",
        "<winPublic>": "%PUBLIC%",
        "<winProgramData>": "%PROGRAMDATA%",
        "<winDir>": "%WINDIR%",
        "<root>": "%INSTALLATION_PATH%",
        "<base>": "%INSTALLATION_PATH%",
        "<storeUserId>": "<user-id>",
    }
    output = path
    for src, dst in mapping.items():
        output = re.sub(re.escape(src), dst, output, flags=re.IGNORECASE)
    return output.replace("/", "\\")


def _is_windows_save_entry(file_meta: dict) -> bool:
    tags = file_meta.get("tags")
    if not isinstance(tags, (list, tuple)):
        return False
    tag_strs = [str(t).strip().lower() for t in tags if t is not None]
    if "save" not in tag_strs:
        return False

    when_obj = file_meta.get("when")
    if when_obj is None:
        return True
    if isinstance(when_obj, list):
        for item in when_obj:
            if not isinstance(item, dict):
                continue
            os_val = item.get("os")
            if os_val is None:
                return True
            os_str = str(os_val).strip().lower()
            if not os_str or os_str == "windows":
                return True
        return False
    return True


def compile_yaml_to_compact_manifest(yaml_text: str) -> dict:
    root = yaml.safe_load(yaml_text)
    if not isinstance(root, dict):
        raise ValueError("manifest root must be a mapping")

    name_index: Dict[str, List[str]] = {}
    steam_index: Dict[str, List[str]] = {}
    aliases: Dict[str, str] = {}
    total_games = 0

    for game_name_raw, entry in root.items():
        game_name = str(game_name_raw or "")
        if not isinstance(entry, dict):
            continue
        total_games += 1

        alias = entry.get("alias")
        if isinstance(alias, str) and alias.strip():
            aliases[game_name] = alias.strip()
            continue

        files = entry.get("files")
        if not isinstance(files, dict):
            continue

        save_paths: List[str] = []
        for fp, fm in files.items():
            path = str(fp or "").strip()
            if not path:
                continue
            if not isinstance(fm, dict) or not _is_windows_save_entry(fm):
                continue
            translated = _translate_manifest_path(path)
            if not any(p.lower() == translated.lower() for p in save_paths):
                save_paths.append(translated)

        if not save_paths:
            continue

        norm_name = normalize_manifest_game_name(game_name)
        if norm_name:
            name_index[norm_name] = list(save_paths)

        steam = entry.get("steam")
        if isinstance(steam, dict):
            sid = str(steam.get("id") or "").strip()
            if sid and sid.isdigit():
                steam_index[sid] = list(save_paths)

    for from_name, to_name in aliases.items():
        src = normalize_manifest_game_name(from_name)
        dst = normalize_manifest_game_name(to_name)
        if dst in name_index:
            name_index[src] = list(name_index[dst])

    return {
        "version": 1,
        "generated_at_unix": int(time.time()),
        "source_url": MANIFEST_URL,
        "stats": {
            "games_total_in_yaml": total_games,
            "games_with_windows_save_paths": len(name_index),
            "steam_ids_indexed": len(steam_index),
        },
        "name_index": name_index,
        "steam_index": steam_index,
    }


class LudusaviManifestProvider:
    def __init__(
        self,
        data_dir: Optional[str] = None,
        bundled_manifest_path: Optional[str] = None,
        http_client: Optional[requests.Session] = None,
    ):
        self._data_dir = data_dir or get_pyqt_user_data_dir()
        self._manifest_path = os.path.join(self._data_dir, MANIFEST_FILENAME)
        self._meta_path = os.path.join(self._data_dir, META_FILENAME)
        self._bundled_manifest_path = bundled_manifest_path or default_bundled_manifest_path()
        self._http = http_client or requests.Session()
        self._lock = threading.Lock()
        self._cache: Optional[dict] = None

    def load_manifest_offline_only(self) -> dict:
        with self._lock:
            if self._cache is not None:
                return self._cache

            doc = self._load_manifest_document_from_disk() or self._seed_manifest_from_bundle()
            if doc is None:
                doc = self._create_empty_manifest()
            self._cache = doc
            return self._cache

    def refresh_now(self) -> str:
        headers: Dict[str, str] = {}
        meta = self._load_meta()
        etag = meta.get("etag", "")
        if etag:
            headers["If-None-Match"] = etag

        try:
            resp = self._http.get(
                MANIFEST_URL,
                headers=headers,
                timeout=MANIFEST_HTTP_TIMEOUT_SEC,
            )
        except requests.RequestException:
            return "network_error"

        if resp.status_code == 304:
            current = self._load_manifest_document_from_disk()
            if current is None:
                return "not_modified_without_cache"
            meta["fetched_at_unix"] = str(int(time.time()))
            self._save_meta(meta)
            with self._lock:
                self._cache = current
            return "not_modified"

        if not resp.ok:
            return f"http_{resp.status_code}"

        if len(resp.content) > MAX_MANIFEST_DOWNLOAD_BYTES:
            return "manifest_too_large"

        try:
            yaml_text = resp.text
        except requests.RequestException:
            return "network_error"

        try:
            compiled = compile_yaml_to_compact_manifest(yaml_text)
        except Exception:
            return "yaml_error"

        self._save_manifest(compiled)
        self._save_meta(
            {
                "fetched_at_unix": str(int(time.time())),
                "etag": resp.headers.get("ETag", ""),
            }
        )
        with self._lock:
            self._cache = compiled
        return "updated"

    def find_save_paths(
        self,
        game_name: str,
        steam_app_id: Optional[str] = None,
        strict_steam_indexing: bool = False,
    ) -> List[str]:
        return self.find_save_paths_with_meta(game_name, steam_app_id, strict_steam_indexing).paths

    def find_save_paths_with_meta(
        self,
        game_name: str,
        steam_app_id: Optional[str] = None,
        strict_steam_indexing: bool = False,
    ) -> LudusaviSaveLookup:
        manifest = self.load_manifest_offline_only()
        app_key = str(steam_app_id or "").strip()

        steam_index = manifest.get("steam_index") or {}
        if app_key and app_key in steam_index:
            steam_entry = steam_index[app_key]
            if isinstance(steam_entry, list):
                paths = [str(x).strip() for x in steam_entry if str(x).strip()]
                return LudusaviSaveLookup(paths, LudusaviMatchKind.STEAM_ID)
            if strict_steam_indexing:
                return LudusaviSaveLookup([], LudusaviMatchKind.STEAM_ID)

        norm = normalize_manifest_game_name(game_name)
        name_index = manifest.get("name_index") or {}
        name_paths = name_index.get(norm)
        if isinstance(name_paths, list):
            paths = [str(x).strip() for x in name_paths if str(x).strip()]
            return LudusaviSaveLookup(paths, LudusaviMatchKind.NAME_INDEX)

        return LudusaviSaveLookup([], LudusaviMatchKind.NONE)

    def _seed_manifest_from_bundle(self) -> Optional[dict]:
        path = self._bundled_manifest_path
        if not path or not os.path.isfile(path):
            return None
        try:
            with open(path, encoding="utf-8") as handle:
                raw = json.load(handle)
            self._save_manifest(raw)
            self._save_meta({"fetched_at_unix": str(int(time.time())), "etag": ""})
            return raw
        except (OSError, json.JSONDecodeError):
            return None

    def _load_manifest_document_from_disk(self) -> Optional[dict]:
        if not os.path.isfile(self._manifest_path):
            return None
        try:
            with open(self._manifest_path, encoding="utf-8") as handle:
                return json.load(handle)
        except (OSError, json.JSONDecodeError):
            return None

    @staticmethod
    def _create_empty_manifest() -> dict:
        return {
            "version": 1,
            "generated_at_unix": 0,
            "source_url": MANIFEST_URL,
            "stats": {},
            "name_index": {},
            "steam_index": {},
        }

    def _load_meta(self) -> Dict[str, str]:
        if not os.path.isfile(self._meta_path):
            return {}
        try:
            with open(self._meta_path, encoding="utf-8") as handle:
                data = json.load(handle)
            return {str(k): str(v) for k, v in data.items()} if isinstance(data, dict) else {}
        except (OSError, json.JSONDecodeError):
            return {}

    def _save_meta(self, meta: Dict[str, str]) -> None:
        _atomic_write_text(self._meta_path, json.dumps(meta, indent=2))

    def _save_manifest(self, manifest: dict) -> None:
        _atomic_write_text(self._manifest_path, json.dumps(manifest, indent=2))

File: src-winui/src/GSBT.Core/Services/LudusaviManifestProvider.cs
using System.Text.Json;
using System.Text.RegularExpressions;
using GSBT.Core.Common;
using GSBT.Core.Models;
using YamlDotNet.Serialization;

namespace GSBT.Core.Services;

public sealed class LudusaviManifestProvider
{
    public const string ManifestUrl = "https://raw.githubusercontent.com/mtkennerly/ludusavi-manifest/master/data/manifest.yaml";
    private const string ManifestFilename = "ludusavi-save-manifest.json";
    private const string MetaFilename = "ludusavi-save-manifest.meta.json";
    private const int MaxManifestDownloadBytes = 64 * 1024 * 1024;
    private const int MaxManifestGames = 100_000;
    private const int MaxPathsPerGame = 128;
    private const int MaxManifestPaths = 500_000;
    private const int MaxManifestPathLength = 4096;
    private static readonly TimeSpan ManifestHttpTimeout = TimeSpan.FromMinutes(3);
    private static readonly Regex NameClean = new("[^a-z0-9]+", RegexOptions.Compiled);

    private readonly string _dataDir;
    private readonly string _manifestPath;
    private readonly string _metaPath;
    private readonly string? _bundledManifestPath;
    private readonly HttpClient _httpClient;
    private readonly object _lock = new();
    private JsonElement? _cache;

    public LudusaviManifestProvider(string? dataDir = null, string? bundledManifestPath = null, HttpClient? httpClient = null)
    {
        _dataDir = dataDir ?? UserDataDir.GetWinUiUserDataDir();
        _manifestPath = Path.Combine(_dataDir, ManifestFilename);
        _metaPath = Path.Combine(_dataDir, MetaFilename);
        _bundledManifestPath = bundledManifestPath;
        _httpClient = httpClient ?? new HttpClient { Timeout = ManifestHttpTimeout };
    }

    public static string NormalizeManifestGameName(string name)
        => string.IsNullOrWhiteSpace(name) ? string.Empty : NameClean.Replace(name.Trim().ToLowerInvariant(), " ").Trim();

    public ManifestProvenance GetProvenance()
    {
        var manifest = LoadManifestOfflineOnly();
        var valid = ValidateCompiledManifest(manifest, out var validationError);
        var meta = LoadMeta();
        var source = meta.TryGetValue("source", out var recordedSource) && !string.IsNullOrWhiteSpace(recordedSource)
            ? recordedSource
            : meta.TryGetValue("etag", out var etag) && !string.IsNullOrWhiteSpace(etag)
                ? "downloaded"
                : File.Exists(_manifestPath)
                    ? "bundled-or-legacy-cache"
                    : "empty";
        var version = manifest.TryGetProperty("version", out var versionElement)
            ? versionElement.ToString()
            : "unknown";
        var generatedAtUtc = TryReadUnixTimestamp(manifest, "generated_at_unix");
        var fetchedAtUtc = meta.TryGetValue("fetched_at_unix", out var fetched)
            && long.TryParse(fetched, out var fetchedUnix)
                ? DateTimeOffset.FromUnixTimeSeconds(fetchedUnix)
                : (DateTimeOffset?)null;
        var sourceUrl = manifest.TryGetProperty("source_url", out var urlElement)
            ? urlElement.GetString()
            : null;
        var sanitizedPathsRemoved = meta.TryGetValue("sanitized_paths_removed", out var removedText)
            && int.TryParse(removedText, out var removed)
                ? removed
                : 0;

        return new ManifestProvenance(
            source,
            version,
            generatedAtUtc,
            fetchedAtUtc,
            valid,
            valid ? "validated" : validationError ?? "validation failed",
            sourceUrl,
            sanitizedPathsRemoved);
    }

    public JsonElement LoadManifestOfflineOnly()
    {
        lock (_lock)
        {
            if (_cache is { } ready)
            {
                return ready;
            }

            var doc = LoadManifestDocumentFromDisk() ?? SeedManifestFromBundle();
            if (doc is null)
            {
                doc = CreateEmptyManifest();
            }

            _cache = doc.Value;
            return _cache.Value;
        }
    }

    public async Task<string> RefreshNowAsync(CancellationToken ct = default)
    {
        lock (_lock)
        {
            // keep sync behavior around shared state
        }

        using var req = new HttpRequestMessage(HttpMethod.Get, ManifestUrl);
        var meta = LoadMeta();
        if (meta.TryGetValue("etag", out var etag) && !string.IsNullOrWhiteSpace(etag))
        {
            req.Headers.TryAddWithoutValidation("If-None-Match", etag);
        }

        HttpResponseMessage? response;
        try
        {
            response = await _httpClient.SendAsync(req, HttpCompletionOption.ResponseHeadersRead, ct);
        }
        catch
        {
            return "network_error";
        }

        using var resp = response;

        if ((int)resp.StatusCode == 304)
        {
            var current = LoadManifestDocumentFromDisk();
            if (current is null)
            {
                return "not_modified_without_cache";
            }

            meta = LoadMeta();
            meta["fetched_at_unix"] = DateTimeOffset.UtcNow.ToUnixTimeSeconds().ToString();
            SaveMeta(meta);
            lock (_lock)
            {
                _cache = current.Value;
            }

            return "not_modified";
        }

        if (!resp.IsSuccessStatusCode)
        {
            return $"http_{(int)resp.StatusCode}";
        }

        string yaml;
        try
        {
            await using var stream = await resp.Content.ReadAsStreamAsync(ct).ConfigureAwait(false);
            using var reader = new StreamReader(stream);
            var sb = new System.Text.StringBuilder();
            var buffer = new char[8192];
            int read;
            while ((read = await reader.ReadAsync(buffer.AsMemory(0, buffer.Length), ct).ConfigureAwait(false)) > 0)
            {
                if (sb.Length + read > MaxManifestDownloadBytes)
                {
                    return "manifest_too_large";
                }

                sb.Append(buffer, 0, read);
            }

            yaml = sb.ToString();
        }
        catch
        {
            return "network_error";
        }

        JsonElement compiled;
        try
        {
            compiled = CompileYamlToCompactManifest(yaml);
            if (!ValidateCompiledManifest(compiled, out _))
            {
                return "manifest_invalid";
            }
        }
        catch
        {
            return "yaml_error";
        }

        SaveManifest(compiled);
        var newMeta = new Dictionary<string, string>
        {
            ["fetched_at_unix"] = DateTimeOffset.UtcNow.ToUnixTimeSeconds().ToString(),
            ["etag"] = resp.Headers.ETag?.Tag ?? string.Empty,
            ["source"] = "downloaded",
            ["sanitized_paths_removed"] = "0",
        };
        SaveMeta(newMeta);

        lock (_lock)
        {
            _cache = compiled;
        }

        return "updated";
    }

    /// <param name="strictSteamIndexing">
    /// When true and <paramref name="steamAppId"/> is set: if the manifest's <c>steam_index</c> contains that id,
    /// use only those paths (no <c>name_index</c> fallback). If that id is <b>not</b> listed in <c>steam_index</c> at all,
    /// fall back to <c>name_index</c> so titles Ludusavi only maps by name still resolve (e.g. some older bundles).
    /// When the id is listed with an empty path list, keep that as "no manifest paths" (no name fallback).
    /// This avoids wrong-title matches when the manifest documents a Steam id that points at a different classic release,
    /// without dropping games that were never given a <c>steam:</c> block for that id.
    /// </param>
    public IReadOnlyList<string> FindSavePaths(string gameName, string? steamAppId, bool strictSteamIndexing = false)
        => FindSavePathsWithMeta(gameName, steamAppId, strictSteamIndexing).Paths;

    /// <inheritdoc cref="FindSavePaths(string, string?, bool)"/>
    /// <returns>Paths plus whether they came from <see cref="LudusaviMatchKind.SteamId"/> or <see cref="LudusaviMatchKind.NameIndex"/>.</returns>
    public LudusaviSaveLookup FindSavePathsWithMeta(string gameName, string? steamAppId, bool strictSteamIndexing = false)
    {
        var manifest = LoadManifestOfflineOnly();
        var appKey = (steamAppId ?? string.Empty).Trim();

        if (manifest.TryGetProperty("steam_index", out var steamIndex) &&
            !string.IsNullOrWhiteSpace(appKey) &&
            steamIndex.TryGetProperty(appKey, out var steamEntryForApp))
        {
            if (steamEntryForApp.ValueKind == JsonValueKind.Array)
            {
                var fromSteam = steamEntryForApp.EnumerateArray().Select(x => x.GetString() ?? string.Empty).Where(x => !string.IsNullOrWhiteSpace(x)).ToList();
                return new LudusaviSaveLookup(fromSteam, LudusaviMatchKind.SteamId);
            }

            // steam_index lists this app id but not as a path array — do not guess from title.
            if (strictSteamIndexing)
            {
                return new LudusaviSaveLookup([], LudusaviMatchKind.SteamId);
            }
        }

        var norm = NormalizeManifestGameName(gameName);
        if (manifest.TryGetProperty("name_index", out var nameIdx) &&
            nameIdx.TryGetProperty(norm, out var namePaths) &&
            namePaths.ValueKind == JsonValueKind.Array)
        {
            var fromName = namePaths.EnumerateArray().Select(x => x.GetString() ?? string.Empty).Where(x => !string.IsNullOrWhiteSpace(x)).ToList();
            return new LudusaviSaveLookup(fromName, LudusaviMatchKind.NameIndex);
        }

        return new LudusaviSaveLookup([], LudusaviMatchKind.None);
    }

    private static JsonElement CompileYamlToCompactManifest(string yaml)
    {
        var deserializer = new DeserializerBuilder().Build();
        var root = deserializer.Deserialize<Dictionary<object, object?>>(yaml);

        var nameIndex = new Dictionary<string, List<string>>(StringComparer.OrdinalIgnoreCase);
        var steamIndex = new Dictionary<string, List<string>>(StringComparer.OrdinalIgnoreCase);
        var aliases = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        var totalGames = 0;
        var totalPaths = 0;

        foreach (var (k, v) in root)
        {
            var gameName = k?.ToString() ?? string.Empty;
            if (v is not Dictionary<object, object?> entry)
            {
                continue;
            }

            totalGames++;
            if (totalGames > MaxManifestGames)
            {
                throw new InvalidDataException($"Manifest contains more than {MaxManifestGames} games.");
            }

            if (entry.TryGetValue("alias", out var aliasObj) && aliasObj is string alias && !string.IsNullOrWhiteSpace(alias))
            {
                aliases[gameName] = alias.Trim();
                continue;
            }

            if (!entry.TryGetValue("files", out var filesObj) || filesObj is not Dictionary<object, object?> files)
            {
                continue;
            }

            var savePaths = new List<string>();
            foreach (var (fp, fm) in files)
            {
                var path = fp?.ToString();
                if (string.IsNullOrWhiteSpace(path))
                {
                    continue;
                }

                if (fm is not Dictionary<object, object?> fileMeta || !IsWindowsSaveEntry(fileMeta))
                {
                    continue;
                }

                var translated = TranslateManifestPath(path.Trim());
                if (!IsSafeManifestPathTemplate(translated))
                {
                    continue;
                }

                if (!savePaths.Contains(translated, StringComparer.OrdinalIgnoreCase))
                {
                    if (savePaths.Count >= MaxPathsPerGame || totalPaths >= MaxManifestPaths)
                    {
                        throw new InvalidDataException("Manifest path limits were exceeded.");
                    }

                    savePaths.Add(translated);
                    totalPaths++;
                }
            }

            if (savePaths.Count == 0)
            {
                continue;
            }

            var normName = NormalizeManifestGameName(gameName);
            if (!string.IsNullOrWhiteSpace(normName))
            {
                nameIndex[normName] = [.. savePaths];
            }

            if (entry.TryGetValue("steam", out var steamObj) && steamObj is Dictionary<object, object?> steamDict &&
                steamDict.TryGetValue("id", out var sidObj))
            {
                var sid = sidObj?.ToString()?.Trim();
                if (!string.IsNullOrWhiteSpace(sid) && sid.All(char.IsDigit))
                {
                    steamIndex[sid] = [.. savePaths];
                }
            }
        }

        foreach (var (fromName, toName) in aliases)
        {
            var src = NormalizeManifestGameName(fromName);
            var dst = NormalizeManifestGameName(toName);
            if (nameIndex.TryGetValue(dst, out var paths))
            {
                nameIndex[src] = [.. paths];
            }
        }

        var payload = new
        {
            version = 1,
            generated_at_unix = DateTimeOffset.UtcNow.ToUnixTimeSeconds(),
            source_url = ManifestUrl,
            stats = new
            {
                games_total_in_yaml = totalGames,
                games_with_windows_save_paths = nameIndex.Count,
                steam_ids_indexed = steamIndex.Count
            },
            name_index = nameIndex,
            steam_index = steamIndex
        };

        return JsonSerializer.SerializeToElement(payload);
    }

    private static bool IsWindowsSaveEntry(Dictionary<object, object?> fileMeta)
    {
        if (!fileMeta.TryGetValue("tags", out var tagsObj) || tagsObj is not IEnumerable<object> tags)
        {
            return false;
        }

        if (!tags.Select(x => x?.ToString()?.Trim().ToLowerInvariant()).Contains("save"))
        {
            return false;
        }

        if (!fileMeta.TryGetValue("when", out var whenObj) || whenObj is null)
        {
            return true;
        }

        if (whenObj is IEnumerable<object> list)
        {
            foreach (var item in list)
            {
                if (item is not Dictionary<object, object?> cond)
                {
                    continue;
                }

                if (!cond.TryGetValue("os", out var osObj))
                {
                    return true;
                }

                var os = osObj?.ToString()?.Trim().ToLowerInvariant();
                if (string.IsNullOrWhiteSpace(os) || os == "windows")
                {
                    return true;
                }
            }

            return false;
        }

        return true;
    }

    private static string TranslateManifestPath(string path)
    {
        var map = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
        {
            ["<home>"] = "~",
            ["<winAppData>"] = "%APPDATA%",
            ["<winLocalAppData>"] = "%LOCALAPPDATA%",
            ["<winLocalAppDataLow>"] = "%USERPROFILE%\\AppData\\LocalLow",
            ["<winDocuments>"] = "%USERPROFILE%\\Documents",
            ["<winPublic>"] = "%PUBLIC%",
            ["<winProgramData>"] = "%PROGRAMDATA%",
            ["<winDir>"] = "%WINDIR%",
            ["<root>"] = "%INSTALLATION_PATH%",
            ["<base>"] = "%INSTALLATION_PATH%",
            ["<storeUserId>"] = "<user-id>"
        };

        var output = path;
        foreach (var (from, to) in map)
        {
            output = output.Replace(from, to, StringComparison.OrdinalIgnoreCase);
        }

        return output.Replace('/', '\\');
    }

    internal static bool IsSafeManifestPathTemplate(string path)
    {
        if (string.IsNullOrWhiteSpace(path) || path.Length > MaxManifestPathLength || path.IndexOf('\0') >= 0)
        {
            return false;
        }

        var normalized = path.Trim().Replace('/', '\\');
        var segments = normalized.Split('\\', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);
        if (segments.Any(static segment => segment == ".."))
        {
            return false;
        }

        string[] unsafeRoots =
        [
            "~",
            "%USERPROFILE%",
            "%APPDATA%",
            "%LOCALAPPDATA%",
            "%PROGRAMDATA%",
            "%PUBLIC%",
            "%WINDIR%",
            "%INSTALLATION_PATH%",
        ];
        if (unsafeRoots.Contains(normalized.TrimEnd('\\'), StringComparer.OrdinalIgnoreCase))
        {
            return false;
        }

        if (normalized.StartsWith("%WINDIR%", StringComparison.OrdinalIgnoreCase))
        {
            return false;
        }

        if (Path.IsPathRooted(normalized)
            && string.Equals(
                Path.GetPathRoot(normalized)?.TrimEnd('\\'),
                normalized.TrimEnd('\\'),
                StringComparison.OrdinalIgnoreCase))
        {
            return false;
        }

        return true;
    }

    private static bool ValidateCompiledManifest(JsonElement manifest, out string? error)
    {
        error = null;
        if (manifest.ValueKind != JsonValueKind.Object
            || !manifest.TryGetProperty("name_index", out var names)
            || names.ValueKind != JsonValueKind.Object
            || !manifest.TryGetProperty("steam_index", out var steam)
            || steam.ValueKind != JsonValueKind.Object)
        {
            error = "Manifest indexes are missing or malformed.";
            return false;
        }

        var games = 0;
        var paths = 0;
        foreach (var index in new[] { names, steam })
        {
            foreach (var property in index.EnumerateObject())
            {
                games++;
                if (games > MaxManifestGames * 2 || property.Value.ValueKind != JsonValueKind.Array)
                {
                    error = "Manifest index limits or shape are invalid.";
                    return false;
                }

                var perGame = 0;
                foreach (var value in property.Value.EnumerateArray())
                {
                    perGame++;
                    paths++;
                    if (value.ValueKind != JsonValueKind.String
                        || !IsSafeManifestPathTemplate(value.GetString() ?? string.Empty)
                        || perGame > MaxPathsPerGame
                        || paths > MaxManifestPaths * 2)
                    {
                        error = "Manifest contains an unsafe path or exceeds path limits.";
                        return false;
                    }
                }
            }
        }

        return true;
    }

    private JsonElement? SeedManifestFromBundle()
    {
        var path = _bundledManifestPath;
        if (string.IsNullOrWhiteSpace(path))
        {
            return null;
        }

        if (!File.Exists(path))
        {
            return null;
        }

        try
        {
            var raw = JsonDocument.Parse(File.ReadAllText(path)).RootElement.Clone();
            if (!TryPrepareManifest(raw, out var prepared, out var removed))
            {
                return null;
            }

            SaveManifest(prepared);
            SaveMeta(new Dictionary<string, string>
            {
                ["fetched_at_unix"] = DateTimeOffset.UtcNow.ToUnixTimeSeconds().ToString(),
                ["etag"] = string.Empty,
                ["source"] = "bundled",
                ["sanitized_paths_removed"] = removed.ToString(),
            });
            return prepared;
        }
        catch
        {
            return null;
        }
    }

    private JsonElement? LoadManifestDocumentFromDisk()
    {
        if (!File.Exists(_manifestPath))
        {
            return null;
        }

        try
        {
            var manifest = JsonDocument.Parse(File.ReadAllText(_manifestPath)).RootElement.Clone();
            if (!TryPrepareManifest(manifest, out var prepared, out var removed))
            {
                return null;
            }

            if (removed > 0)
            {
                SaveManifest(prepared);
                var meta = LoadMeta();
                meta["sanitized_paths_removed"] = removed.ToString();
                SaveMeta(meta);
            }

            return prepared;
        }
        catch
        {
            return null;
        }
    }

    private static JsonElement CreateEmptyManifest() => JsonSerializer.SerializeToElement(new
    {
        version = 1,
        generated_at_unix = 0,
        source_url = ManifestUrl,
        stats = new { },
        name_index = new Dictionary<string, string[]>(),
        steam_index = new Dictionary<string, string[]>()
    });

    private static bool TryPrepareManifest(
        JsonElement manifest,
        out JsonElement prepared,
        out int removedPaths)
    {
        prepared = default;
        removedPaths = 0;
        if (ValidateCompiledManifest(manifest, out _))
        {
            prepared = manifest;
            return true;
        }

        if (manifest.ValueKind != JsonValueKind.Object
            || !manifest.TryGetProperty("name_index", out var names)
            || names.ValueKind != JsonValueKind.Object
            || !manifest.TryGetProperty("steam_index", out var steam)
            || steam.ValueKind != JsonValueKind.Object)
        {
            return false;
        }

        var sanitizedNames = SanitizeIndex(names, ref removedPaths);
        var sanitizedSteam = SanitizeIndex(steam, ref removedPaths);
        if (sanitizedNames is null || sanitizedSteam is null)
        {
            return false;
        }

        var payload = new Dictionary<string, object?>
        {
            ["version"] = manifest.TryGetProperty("version", out var version) ? version.Clone() : 1,
            ["generated_at_unix"] = manifest.TryGetProperty("generated_at_unix", out var generated) ? generated.Clone() : 0,
            ["source_url"] = manifest.TryGetProperty("source_url", out var sourceUrl) ? sourceUrl.Clone() : ManifestUrl,
            ["stats"] = manifest.TryGetProperty("stats", out var stats) ? stats.Clone() : new { },
            ["name_index"] = sanitizedNames,
            ["steam_index"] = sanitizedSteam,
        };
        prepared = JsonSerializer.SerializeToElement(payload);
        return removedPaths > 0 && ValidateCompiledManifest(prepared, out _);
    }

    private static Dictionary<string, string[]>? SanitizeIndex(JsonElement index, ref int removedPaths)
    {
        var result = new Dictionary<string, string[]>(StringComparer.OrdinalIgnoreCase);
        var entries = 0;
        var paths = 0;
        foreach (var property in index.EnumerateObject())
        {
            entries++;
            if (entries > MaxManifestGames * 2 || property.Value.ValueKind != JsonValueKind.Array)
            {
                return null;
            }

            var safe = property.Value.EnumerateArray()
                .Where(static value => value.ValueKind == JsonValueKind.String)
                .Select(static value => value.GetString() ?? string.Empty)
                .Where(IsSafeManifestPathTemplate)
                .Distinct(StringComparer.OrdinalIgnoreCase)
                .Take(MaxPathsPerGame + 1)
                .ToArray();
            var originalCount = property.Value.GetArrayLength();
            removedPaths += originalCount - safe.Length;
            paths += safe.Length;
            if (safe.Length > MaxPathsPerGame || paths > MaxManifestPaths * 2)
            {
                return null;
            }

            if (safe.Length > 0)
            {
                result[property.Name] = safe;
            }
        }

        return result;
    }

    private static DateTimeOffset? TryReadUnixTimestamp(JsonElement manifest, string propertyName)
    {
        if (!manifest.TryGetProperty(propertyName, out var value)
            || !value.TryGetInt64(out var unix)
            || unix <= 0)
        {
            return null;
        }

        try
        {
            return DateTimeOffset.FromUnixTimeSeconds(unix);
        }
        catch (ArgumentOutOfRangeException)
        {
            return null;
        }
    }

    private Dictionary<string, string> LoadMeta()
    {
        if (!File.Exists(_metaPath))
        {
            return new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        }

        try
        {
            return JsonSerializer.Deserialize<Dictionary<string, string>>(File.ReadAllText(_metaPath))
                   ?? new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        }
        catch
        {
            return new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        }
    }

    private void SaveMeta(Dictionary<string, string> meta)
    {
        Directory.CreateDirectory(_dataDir);
        AtomicFileWrite.WriteAllText(_metaPath, JsonSerializer.Serialize(meta, new JsonSerializerOptions { WriteIndented = true }));
    }

    private void SaveManifest(JsonElement manifest)
    {
        if (!ValidateCompiledManifest(manifest, out var error))
        {
            throw new InvalidDataException(error ?? "Manifest validation failed.");
        }

        Directory.CreateDirectory(_dataDir);
        AtomicFileWrite.WriteAllText(_manifestPath, JsonSerializer.Serialize(manifest, new JsonSerializerOptions { WriteIndented = true }));
    }
}

public sealed record ManifestProvenance(
    string Source,
    string Version,
    DateTimeOffset? GeneratedAtUtc,
    DateTimeOffset? FetchedAtUtc,
    bool IsValid,
    string ValidationStatus,
    string? SourceUrl,
    int SanitizedPathsRemoved);
